### ISSUE TITLE
Model functions: parameter vector as first or second argument?

### ISSUE THREAD

**Issue by stestoll** [luisfabib/DeerAnalysis2#108]
Currently, the parameter vector is the second argument for model functions: `dd_gauss(r,par)`. We should consider whether there are advantages of putting it first, i.e. `dd_gauss(par,r)`. It is not clear what's better.

Putting the parameter first has the advantage that users can define model functions that do not depend on anything except a parameter vector: `mymodel(par)`.

There is no obvious downside (beyond that it is a breaking change).

Least-squares fitting interfaces in other languages have both orderings:
- Parameters first:
  - [`lsqcurvefit`](https://www.mathworks.com/help/optim/ug/lsqcurvefit.html) from the MATLAB optimization toolbox uses `model(par,x)`
  - [`lsqnonlin`](https://www.mathworks.com/help/optim/ug/lsqnonlin.html) from the MATLAB optimization toolbox uses `residualvec(par)`. Extra parameters have to be passed via anonymous function wrappers etc.
  - [`scipy.optimize.least_squares`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html#scipy.optimize.least_squares) (Python) uses `residualvec(par,...)`
- Independent variable first
  - [`scipy.optimize.curve_fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html) (Python) uses `model(x,par1,par2,...)`
  - [LMFIT](https://lmfit.github.io/lmfit-py/model.html#the-model-class), another Python least-squares package, requires parameters as first input when evaluating a model (`y = model.eval(params, x=x)`, but the independent variable as first input when fitting the model (`result = gmodel.fit(y, params, x=x)` or `result = gmodel.fit(y, x=x, cen=6.5, amp=100, wid=2.0)`)
  - [`LsqFit.jl`](https://julianlsolvers.github.io/LsqFit.jl/latest/api/) (Julia) uses `model(x,par)` and does the fitting with `curve_fit(model, x, y, p0)`

There is also the [`nls`](https://www.rdocumentation.org/packages/stats/topics/nls) function in the [R stats package](https://www.rdocumentation.org/packages/stats), several functions in [MINPACK](https://www.netlib.org/minpack/)

There is a distinction between curve fitting functions and general non-linear least-squares. The former require a model function, whereas the latter require a function that returns the residual vector.

The general pattern seems to be that curve fitting model functions take the independent variable first (except in MATLAB), and residual-vector functions for general least-squares take the parameters first.
